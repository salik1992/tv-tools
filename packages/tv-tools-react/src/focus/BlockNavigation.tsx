import type { PropsWithChildren } from 'react';
import { FocusContext, useFocusContainer } from '../focus';

function markAsProcessed() {
	return true;
}

/**
 * BlockNavigation is a component that prevents navigation to escape from its children.
 * Useful for Modal, VirtualKeyboard, etc.
 */
export const BlockNavigation = ({ children }: PropsWithChildren) => {
	const { focusContextValue, useOnLeft, useOnUp, useOnDown, useOnRight } =
		useFocusContainer();

	useOnLeft(markAsProcessed);
	useOnRight(markAsProcessed);
	useOnUp(markAsProcessed);
	useOnDown(markAsProcessed);

	return (
		<FocusContext.Provider value={focusContextValue}>
			{children}
		</FocusContext.Provider>
	);
};
