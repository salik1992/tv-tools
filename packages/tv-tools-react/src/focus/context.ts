import { createContext } from 'react';
import { noop } from '@salik1992/tv-tools/utils/noop';
import type { Focus } from './types';

export const FocusContext = createContext<Focus>({
	addChild: noop,
});
