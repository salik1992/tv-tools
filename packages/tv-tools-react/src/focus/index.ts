export { FocusRoot } from './FocusRoot';
export { HorizontalFocus } from './HorizontalFocus';
export { Interactable } from './Interactable';
export { VerticalFocus } from './VerticalFocus';
export { useFocusContainer } from './useFocusContainer';
export { useKeySequenceListener } from './useKeySequenceListener';
