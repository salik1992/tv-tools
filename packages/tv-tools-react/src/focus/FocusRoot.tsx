import type { FocusEvent, KeyboardEvent, PropsWithChildren } from 'react';
import { useCallback } from 'react';
import { isDirectional } from '@salik1992/tv-tools/control';
import { focus } from '@salik1992/tv-tools/focus';

/**
 * The main root component for the focus management.
 * It creates a div element that is used for listening to key events in the manager.
 * Use at the root of your application.
 */
export const FocusRoot = ({
	children,
	/** Use this to prevent TV use their own navigation. */
	alwaysPreventNavigationalEvents,
}: PropsWithChildren<{
	alwaysPreventNavigationalEvents?: boolean;
}>) => {
	const onFocus = useCallback((event: FocusEvent) => {
		focus.handleFocusEvent(event.nativeEvent);
	}, []);

	const keyDownCapture = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keydown', 'capture', event);
	}, []);

	const keyDownBubble = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keydown', 'bubble', event);
		if (
			alwaysPreventNavigationalEvents &&
			isDirectional(event) &&
			document.activeElement?.tagName !== 'INPUT'
		) {
			event.preventDefault();
			event.stopPropagation();
		}
	}, []);

	const keyUpCapture = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keyup', 'capture', event);
	}, []);

	const keyUpBubble = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keyup', 'bubble', event);
		if (
			alwaysPreventNavigationalEvents &&
			isDirectional(event) &&
			document.activeElement?.tagName !== 'INPUT'
		) {
			event.preventDefault();
			event.stopPropagation();
		}
	}, []);

	return (
		<div
			id="focusRoot"
			onFocus={onFocus}
			onKeyDownCapture={keyDownCapture}
			onKeyDown={keyDownBubble}
			onKeyUpCapture={keyUpCapture}
			onKeyUp={keyUpBubble}
		>
			{children}
		</div>
	);
};
