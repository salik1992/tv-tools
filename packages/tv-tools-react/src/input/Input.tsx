import {
	type InputHTMLAttributes,
	useEffect,
	useMemo,
	useRef,
	useState,
} from 'react';
import { Interactable as InteractableBase } from '@salik1992/tv-tools/focus';
import {
	type CharRenderData,
	Input as InputBase,
	type PxRenderData,
	charRenderDataToPxRenderData,
} from '@salik1992/tv-tools/input';
import { Interactable } from '../focus';

/**
 * Input component that wraps the Interactable and Input classes from tv-tools.
 *
 * @prop id - optional id that will be assigned to div and also used for focus management.
 * @prop className - optional class name for the wrapping element.
 * @prop style - optional style for the wrapping element.
 * @prop tabIndex - optional tab index for the Interactable component.
 * @prop focusOnMount - optional mark to tell the component to focus itself when mounted.
 * @prop disabled - optional flag to disable the input.
 * @prop other props - other props that will be passed to the input element.
 */
export const Input = ({
	id: passedId,
	className,
	style,
	tabIndex,
	focusOnMount,
	disabled = false,
	...props
}: {
	focusOnMount?: boolean;
} & InputHTMLAttributes<HTMLInputElement>) => {
	// Instance of Interactable and Input classes
	const interactable = useMemo(
		() => (disabled ? null : new InteractableBase(passedId, tabIndex)),
		[passedId, disabled],
	);
	const input = useMemo(
		() => (!interactable ? null : new InputBase(interactable)),
		[interactable],
	);

	// References to the input and text elements
	const inputRef = useRef<HTMLInputElement>(null);
	const textRef = useRef<HTMLDivElement>(null);

	// Render data handling
	const [renderData, setRenderData] = useState<PxRenderData>(
		charRenderDataToPxRenderData(
			input?.getRenderData(),
			textRef.current?.firstChild,
		),
	);

	const onRenderData = (renderData: CharRenderData) => {
		setRenderData(
			charRenderDataToPxRenderData(
				renderData,
				textRef.current?.firstChild,
			),
		);
	};

	useEffect(() => {
		if (input) {
			input.addEventListener('renderData', onRenderData);
		}
		return () => {
			if (input) {
				input.removeEventListener('renderData', onRenderData);
			}
		};
	}, [input]);

	// Pass element to the tv-tools Input class
	useEffect(() => {
		input?.setInputElement(inputRef.current);
	}, [inputRef.current]);

	// Text, caret and selection rendering
	const visualText = (
		<>
			<div className="text" ref={textRef}>
				{renderData.value}
			</div>
			<div
				className="caret"
				style={{ transform: `translateX(${-renderData.caret}px)` }}
			/>
			{renderData.selection && (
				<div
					className="selection"
					style={{
						transform: `translateX(${-renderData.selection[0]}px)`,
						width: renderData.selection[1],
					}}
				/>
			)}
		</>
	);

	// If there is no Interactable (disabled), return the visual text only.
	if (!interactable) {
		return (
			<div className={className} style={style}>
				{visualText}
			</div>
		);
	}

	let fullClassName = className ?? '';
	if (renderData.active) {
		fullClassName += ' active';
	}
	if (renderData.placeholder) {
		fullClassName += ' placeholder';
	}

	return (
		<Interactable
			interactableBase={interactable}
			className={fullClassName}
			style={style}
			focusOnMount={focusOnMount}
		>
			{visualText}
			<input
				{...props}
				id={`${interactable.id}-input`}
				tabIndex={tabIndex}
				ref={inputRef}
			/>
		</Interactable>
	);
};
