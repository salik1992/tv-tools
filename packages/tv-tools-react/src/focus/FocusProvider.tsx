import type { ComponentProps, FocusEvent, KeyboardEvent } from 'react';
import { useCallback, useMemo } from 'react';
import { isDirectional } from '@salik1992/tv-tools/control';
import { FocusManager } from '@salik1992/tv-tools/focus';
import { noop } from '@salik1992/tv-tools/utils/noop';
import { FocusContext } from './context';

/**
 * The main root component for the focus management.
 * It creates a div element that is used for listening to key events in the manager and
 * it also creates the focus manager instance for the application which is available through
 * the FocusContext or useFocusManager hook.
 * Use at the root of your application.
 * @prop alwaysPreventNavigationalEvents - Use this to prevent TV use their own navigation.
 * @prop ...props - Any other props that you want to pass to the wrapping div element.
 */
export const FocusProvider = ({
	children,
	alwaysPreventNavigationalEvents,
	onFocus: onOuterFocus,
	onKeyDownCapture: onOuterKeyDownCapture,
	onKeyDown: onOuterKeyDown,
	onKeyUpCapture: onOuterKeyUpCapture,
	onKeyUp: onOuterKeyUp,
	...props
}: ComponentProps<'div'> & {
	/** Use this to prevent TV use their own navigation. */
	alwaysPreventNavigationalEvents?: boolean;
}) => {
	const focusManager = useMemo(() => new FocusManager(), []);

	const onFocus = useCallback(
		(event: FocusEvent<HTMLDivElement>) => {
			focusManager.handleFocusEvent(event.nativeEvent);
			onOuterFocus?.(event);
		},
		[focusManager, onOuterFocus],
	);

	const keyDownCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'capture', event);
			onOuterKeyDownCapture?.(event);
		},
		[focusManager, onOuterKeyDownCapture],
	);

	const keyDownBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyDown?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyDown],
	);

	const keyUpCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'capture', event);
			onOuterKeyUpCapture?.(event);
		},
		[focusManager, onOuterKeyUpCapture],
	);

	const keyUpBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyUp?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyUp],
	);

	const focusContextValue = useMemo(
		() => ({
			focusManager,
			addChild: noop,
		}),
		[focusManager],
	);

	return (
		<div
			{...props}
			onFocus={onFocus}
			onKeyDownCapture={keyDownCapture}
			onKeyDown={keyDownBubble}
			onKeyUpCapture={keyUpCapture}
			onKeyUp={keyUpBubble}
		>
			<FocusContext.Provider value={focusContextValue}>
				{children}
			</FocusContext.Provider>
		</div>
	);
};
