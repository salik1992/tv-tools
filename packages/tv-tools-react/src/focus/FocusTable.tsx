import {
	type ComponentProps,
	type ComponentType,
	type PropsWithChildren,
	type ReactNode,
	createContext,
	useCallback,
	useContext,
	useEffect,
	useMemo,
} from 'react';
import type { TableFocusContainer } from '@salik1992/tv-tools/focus';
import { FocusContext } from './context';
import { useTableFocusContainer } from './useTableFocusContainer';

const ChildrenOnlyComponent = ({ children }: PropsWithChildren) => {
	return <>{children}</>;
};

const TableContext = createContext<{
	container?: TableFocusContainer;
	TrComponent: ComponentType<PropsWithChildren<unknown>>;
	TdComponent: ComponentType<PropsWithChildren<unknown>>;
}>({
	TrComponent: ChildrenOnlyComponent,
	TdComponent: ChildrenOnlyComponent,
});

const FocusTd = ({
	rowSpan,
	colSpan,
	children,
	...props
}: PropsWithChildren<{
	colSpan?: number;
	rowSpan?: number;
}>) => {
	const { addChild, ...originalfocusContextValue } = useContext(FocusContext);
	const { TdComponent } = useContext(TableContext);

	const addChildWithSpans = useCallback(
		(childId: string) => {
			addChild(childId, { rowSpan, colSpan });
		},
		[addChild, rowSpan, colSpan],
	);

	const focusContextValue = useMemo(
		() => ({ ...originalfocusContextValue, addChild: addChildWithSpans }),
		[addChildWithSpans, originalfocusContextValue],
	);

	return (
		<FocusContext.Provider value={focusContextValue}>
			<TdComponent {...props}>{children}</TdComponent>
		</FocusContext.Provider>
	);
};

const FocusTr = (props: PropsWithChildren<unknown>) => {
	const { TrComponent, container } = useContext(TableContext);

	useEffect(() => {
		// Needs to run every time to maintain children order.
		container?.addRow();
	});

	return <TrComponent {...props} />;
};

export type FocusTableRenderComponents<
	TrComponent extends ComponentType,
	TdComponent extends ComponentType,
> = {
	FocusTr: (props: ComponentProps<TrComponent>) => ReactNode;
	FocusTd: (
		props: {
			colSpan?: number;
			rowSpan?: number;
		} & ComponentProps<TdComponent>,
	) => ReactNode;
};

export const FocusTable = ({
	id,
	children,
	// @ts-expect-error: TODO
	TableComponent = ChildrenOnlyComponent,
	TrComponent = ChildrenOnlyComponent,
	TdComponent = ChildrenOnlyComponent,
	...tableProps
}: {
	children: (
		components: FocusTableRenderComponents<
			typeof TrComponent,
			typeof TdComponent
		>,
		container: TableFocusContainer,
	) => ReactNode;
	id?: string;
	TableComponent?: ComponentType;
	TrComponent?: ComponentType;
	TdComponent?: ComponentType;
} & ComponentProps<typeof TableComponent>) => {
	const {
		focusContextValue,
		container,
		useOnLeft,
		useOnUp,
		useOnDown,
		useOnRight,
	} = useTableFocusContainer(id);

	useOnLeft(
		(event) =>
			container.moveFocus({ x: -1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnRight(
		(event) =>
			container.moveFocus({ x: 1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnUp(
		(event) =>
			container.moveFocus({ y: -1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnDown(
		(event) =>
			container.moveFocus({ y: 1 }, (event.target as HTMLElement).id),
		[container],
	);

	const tableContextValue = useMemo(
		() => ({
			container,
			TrComponent,
			TdComponent,
		}),
		[TrComponent, TdComponent],
	);

	return (
		<TableContext.Provider value={tableContextValue}>
			<FocusContext.Provider value={focusContextValue}>
				<TableComponent {...tableProps}>
					{children({ FocusTr, FocusTd }, container)}
				</TableComponent>
			</FocusContext.Provider>
		</TableContext.Provider>
	);
};
