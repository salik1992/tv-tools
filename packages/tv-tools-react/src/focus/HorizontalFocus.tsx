import type { KeyboardEvent, PropsWithChildren } from 'react';
import { useFocusContainer } from './useFocusContainer';
import { FocusContext } from './context';

/**
 * Component for markup of where the focus should navigate horizontally.
 * Direct focusable children will be navigated with left/right arrows in the order
 * they are placed in the DOM structure. This component does not render any
 * element. By default this component changes direction when the document is in the
 * right to left layout.
 *
 * @prop id - optional id if you need to address it manually
 * @prop ignoreRtl - optional override to keep the same direction in ltr/rtl layouts
 *
 * @example
 * ```typescriptreact
 * export const Menu = ({ ... }) => {
 *     // ...
 *     return (
 *          <div id="top-menu">
 *              <HorizontalFocus>
 *                  <MenuItem to="home">Home</MenuItem>
 *                  <MenuItem to="search">Search</MenuItem>
 *                  <MenuItem to="epg">EPG</MenuItem>
 *                  <MenuItem to="profile">My Profile</MenuItem>
 *              </HorizontalFocus>
 *          </div>
 *     )
 * }
 * ```
 * If the MenuItem are focusable (using Interactable component internally),
 * then when they are focused and we press left/right, the focus will move between
 * them in the order home <-> search <-> epg <-> profile.
 */
export const HorizontalFocus = ({
	id,
	ignoreRtl,
	children,
}: PropsWithChildren<{ id?: string; ignoreRtl?: boolean }>) => {
	const { focusContextValue, container, useOnLeft, useOnRight } =
		useFocusContainer(id);

	useOnLeft(
		// @ts-ignore: TODO reversed limitation
		(event: KeyboardEvent<HTMLElement>) => {
			return container.moveFocus(-1, (event.target as HTMLElement).id);
		},
		[container],
		{ ignoreRtl },
	);

	useOnRight(
		// @ts-ignore: TODO reversed limitation
		(event: KeyboardEvent<HTMLElement>) => {
			return container.moveFocus(1, (event.target as HTMLElement).id);
		},
		[container],
		{ ignoreRtl },
	);

	return (
		<FocusContext.Provider value={focusContextValue}>
			{children}
		</FocusContext.Provider>
	);
};
