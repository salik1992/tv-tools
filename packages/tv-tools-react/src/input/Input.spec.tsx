import { act, render } from '@testing-library/react';
import { FocusProvider } from '../focus';
import { Input } from './Input';

jest.useFakeTimers();

describe('Input', () => {
	it('should render the Input with Interactable', () => {
		const { container, unmount } = render(
			<Input
				id="input-1"
				type="text"
				value={'test'}
				className="class-1"
				style={{ borderColor: 'black ' }}
				readOnly
			/>,
			{ wrapper: FocusProvider },
		);
		expect(container.innerHTML).toMatchSnapshot();
		unmount();
	});

	it('should render the Input without Interactable and input when disabled', () => {
		const { container, unmount } = render(
			<Input
				id="input-1"
				type="text"
				value={'test'}
				className="class-1"
				style={{ borderColor: 'black ' }}
				readOnly
				disabled
			/>,
			{ wrapper: FocusProvider },
		);
		expect(container.innerHTML).toMatchSnapshot();
		unmount();
	});

	it('should update render data and run proper callbacks when input', () => {
		const onChange = jest.fn();
		const { container, unmount } = render(
			<Input
				id="input-1"
				type="text"
				value={'test'}
				className="class-1"
				style={{ borderColor: 'black ' }}
				onChange={onChange}
			/>,
			{ wrapper: FocusProvider },
		);
		const input = container.querySelector('input');
		if (!input) {
			throw new Error('Input not found');
		}
		act(() => {
			input.value = 'test-changed';
			input.dispatchEvent(new InputEvent('change'));
			jest.runAllTimers();
		});
		expect(container.innerHTML).toMatchSnapshot();
		unmount();
	});
});
