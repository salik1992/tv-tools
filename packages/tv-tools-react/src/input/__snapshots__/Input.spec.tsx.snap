// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render the Input with Interactable 1`] = `"<div class="class-1" style="border-color: black;" id="input-1" tabindex="0"><div class="text"></div><div class="caret" style="transform: translateX(0px);"></div><input readonly="" id="input-1-input" type="text" value="test"></div>"`;

exports[`Input should render the Input without Interactable and input when disabled 1`] = `"<div class="class-1" style="border-color: black;"><div class="text"></div><div class="caret" style="transform: translateX(0px);"></div></div>"`;

exports[`Input should update render data and run proper callbacks when input 1`] = `"<div class="class-1" style="border-color: black;" id="input-1" tabindex="0"><div class="text">test-changed</div><div class="caret" style="transform: translateX(-12px);"></div><input id="input-1-input" type="text" value="test"></div>"`;
